include/mqtt/callable_overlay.hpp

```diff
@@ -726,16 +731,14 @@ struct callable_overlay final : public Impl
     /**
      * @brief next read handler
      *        This handler is called when the current mqtt message has been processed.
+     *        This calls base::on_mqtt_message_processed() only if mqtt_message_processed handler is not set.
      * @param func A callback function that is called when async operation will finish.
      */
-    MQTT_ALWAYS_INLINE void on_mqtt_message_processed(MQTT_NS::any session_life_keeper) noexcept override final
-    {
-        if(h_mqtt_message_processed_)
-        {
+    MQTT_ALWAYS_INLINE void on_mqtt_message_processed(MQTT_NS::any session_life_keeper) noexcept override final {
+        if(h_mqtt_message_processed_) {
```

Comment:
This is the only place that you're not always calling the base class implementation before calling the provided callback.

Should we make them all behave the same?

I think either always call base, then callback,

or always call callback and not base if callback exists, and require end-user to call base implementation if they want it to be called.

one of those.

Reply:
It very difficult.
The current situation is,
* `client` inherits `endpoint` and override `on_pre_send()`, `on_close()`, and `on_error()`. And implement important process *1 in the override definition (timer reset, timer cancel).
* `client` has `set_pre_send_handler()`, `set_on_close()`, and `set_on_error()`. If a user of `client` call them, and the event happens, set handler should be called. Either the user set handlers or not, *1 should be executed automatically.
* `endpoint` has the default definition of `on_mqtt_message_processed()`. `client` doesn't override `on_mqtt_message_processed()`. If a user override this, then the default definition shouldn't be called. It makes conflict behavior.

So the current implementation is correct.

Reply:
Ok.

Can you add those comments to the PR?

Reply:
Ok.

Reply:
Do you mean in the code? Where do you think good?

Reply:
In the documentation for `on_mqtt_message_processed()`, because it is special-case compared to the others.

Reply:
Ok